### ISSUE TITLE
support writing tests in python

### ISSUE THREAD

**Issue by DetachHead** [robotframework/robotframework#4834]
it would be nice if you could write robot tests in python, like you can with keywords.

for example, the following robot file:
```robot
*** Settings ***
Test Setup          ...
Test Teardown       ...


*** Test Cases ***
Test case 1
    Some Keyword

Test case 2
    Some Keyword


*** Keywords ***
Some Keyword
    No Operation
```

could be written in python like:
```py
from typing import override
from robot import TestSuite, test_case


class Foo(TestSuite):
    @override
    def test_setup(self):
        ...

    @override
    def test_teardown(self):
        ...

    @test_case
    def test_case_1(self):
        self.some_keyword()

    @test_case
    def test_case_2(self):
        self.some_keyword()

    @keyword
    def some_keyword(self):
        ...
```

**Comment by DetachHead** [comment]
> Good news is that Robot's execution model ought to already now make it possible to create an external tool that does this. It would allow experimenting with different APIs without being tied to Robot's release schedule and strict backwards compatibility. Anyone interested to experiment?

i'd be willing to give it a shot. would you be able to point me in the right direction?

**Comment by DetachHead** [comment]
@pekkaklarck i gave it a go and made a pytest plugin https://github.com/DetachHead/pytest-robotframework

it's very early stages and there are probably lots of issues with it atm like how it runs the tests twice (first in pytest, then in robot). if you want to take a look and provide any feedback that would be appreciated

**Comment by DetachHead** [comment]
the plugin is pretty much done, however there are a few hacky parts that i could use some guidance on.

### dynamic keywords

i dynamically create functions and give them unique names so they can be discovered as keywords by robot (like `pytestrobotkeyword_12341234_function_name`)

https://github.com/DetachHead/pytest-robotframework/blob/77af729e74859166be9e08c08b7569572852ecfc/pytest_robotframework/_common.py#L84-L97

is there a way to customize the keyword discovery so that this isn't required? the downside of this is that the `xml` file still has the gross unique name because it gets generated before my `prerebotmodifier` changes the names back:

https://github.com/DetachHead/pytest-robotframework/blob/77af729e74859166be9e08c08b7569572852ecfc/pytest_robotframework/_common.py#L65-L76

### replacing test case `body` with wrapped function in a prerunmodifier

i also added support for running `.robot` files so we could gradually adopt the plugin in our codebase while still running all of our tests with the same tool.

but since my plugin injects a dynamically defined "run test" keyword into the existing test case, the original test case body has to be sent to the pytest item's `runtest` function which then executes it with a `BodyRunner`, but i'm not sure if this is the best way to go about it.

https://github.com/DetachHead/pytest-robotframework/blob/77af729e74859166be9e08c08b7569572852ecfc/pytest_robotframework/_common.py#L187-L190

https://github.com/DetachHead/pytest-robotframework/blob/77af729e74859166be9e08c08b7569572852ecfc/pytest_robotframework/_robot.py#L96-L100

### `__init__` files

i'm not sure if i'm handling `__init__.py`/`__init__.robot` files correctly. how should `parse_init` behave differently to `parse` in a custom parser?

https://github.com/DetachHead/pytest-robotframework/blob/77af729e74859166be9e08c08b7569572852ecfc/pytest_robotframework/_python.py#L72-L76

what i have there forces me to enable `runemptysuite`:

https://github.com/DetachHead/pytest-robotframework/blob/77af729e74859166be9e08c08b7569572852ecfc/pytest_robotframework/pytest_robotframework.py#L59

**Comment by pekkaklarck** [comment]
Great work with the plugin! I'm somewhat slowly returning from holidays and need to start discussion about RF 7, so I won't have time to help too much, but I highly recommend you to announce this so that others interested in could participate. I try to help with questions related to Robot's APIs, though. Below are quick answers to your questions:

1. Creating keywords dynamically should probably get a new API. Does #4252 look like something that would work in this case?
2. `parse_init` should return a root suite for the suites in that directory. I can also optionally configure certain settings that can be specified also in init files.

For further discussion it would probably be best to use our Slack. Because support to create tests directly in Python will at least initially be implemented externally, we probably should close this issue. Separate issues can be submitted if there are needs for new APIs used by such tools.

**Comment by DetachHead** [comment]
> 1. Creating keywords dynamically should probably get a new API. Does https://github.com/robotframework/robotframework/issues/4252 look like something that would work in this case?

i don't think so, however it would definitely solve the other problem i had where keywords don't show in the log when called from python tests (i created a custom `keyword` decorator [here](https://github.com/DetachHead/pytest-robotframework/blob/77af729e74859166be9e08c08b7569572852ecfc/pytest_robotframework/__init__.py#L49-L64) for users of the plugin as a workaround)

the issue i was talking about is different though. it's that i don't think there's a way to specify a function reference inside a `running.TestCase`. since i use a custom parser to build it, i have to way to reference my dynamic functions at that stage, so i instead have to add it to a fake module with a unique name and import it using `running.TestSuite.resource.imports.library`.

**Comment by DetachHead** [comment]
so it turns out i didn't really need to dynamically create keywords like that, and i reworked it to just use a regular robot library (https://github.com/DetachHead/pytest-robotframework/pull/57)

i think the main problems i'm facing now is #4252, which i'm trying to address in https://github.com/DetachHead/pytest-robotframework/pull/66 by patching both pytest functions and robot builtin keywords to make them appear in the log
